File: frontend/lib/websocket.ts
/**
 * WebSocket client for real-time communication
 */

import { io, Socket } from 'socket.io-client';

const WS_URL = process.env.NEXT_PUBLIC_WS_URL || 'ws://localhost:5000';

type EventCallback = (data: any) => void;

class WebSocketClient {
  private socket: Socket | null = null;
  private connected: boolean = false;
  private eventCallbacks: Map<string, Set<EventCallback>> = new Map();

  connect(token: string): void {
    if (this.connected) return;

    this.socket = io(WS_URL, {
      auth: { token },
      transports: ['websocket'],
      reconnection: true,
      reconnectionDelay: 1000,
      reconnectionDelayMax: 5000,
      reconnectionAttempts: 5,
    });

    this.socket.on('connect', () => {
      console.log('WebSocket connected');
      this.connected = true;
      this.emit('connection_status', { connected: true });
    });

    this.socket.on('disconnect', () => {
      console.log('WebSocket disconnected');
      this.connected = false;
      this.emit('connection_status', { connected: false });
    });

    this.socket.on('error', (error) => {
      console.error('WebSocket error:', error);
      this.emit('error', error);
    });

    // Register event listeners
    this.setupEventListeners();
  }

  disconnect(): void {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
      this.connected = false;
    }
  }

  isConnected(): boolean {
    return this.connected;
  }

  private setupEventListeners(): void {
    if (!this.socket) return;

    // Agent events
    this.socket.on('agent_typing', (data) => this.emit('agent_typing', data));
    this.socket.on('agent_response_chunk', (data) => this.emit('agent_response_chunk', data));
    this.socket.on('agent_response_complete', (data) => this.emit('agent_response_complete', data));
    this.socket.on('agent_error', (data) => this.emit('agent_error', data));
    this.socket.on('agent_interrupted', (data) => this.emit('agent_interrupted', data));

    // User events
    this.socket.on('user_typing', (data) => this.emit('user_typing', data));

    // Session events
    this.socket.on('joined_session', (data) => this.emit('joined_session', data));
    this.socket.on('left_session', (data) => this.emit('left_session', data));

    // Ping/pong
    this.socket.on('pong', (data) => this.emit('pong', data));
  }

  // Session management
  joinSession(sessionId: string, token: string): void {
    if (!this.socket) return;
    this.socket.emit('join_session', { session_id: sessionId, token });
  }

  leaveSession(sessionId: string): void {
    if (!this.socket) return;
    this.socket.emit('leave_session', { session_id: sessionId });
  }

  // Send message to agent
  sendMessage(
    sessionId: string,
    agentId: string,
    message: string,
    context?: any,
    token?: string
  ): void {
    if (!this.socket) return;
    this.socket.emit('agent_message', {
      session_id: sessionId,
      agent_id: agentId,
      message,
      context,
      token,
    });
  }

  // Interrupt agent
  interruptAgent(sessionId: string, agentId: string, token?: string): void {
    if (!this.socket) return;
    this.socket.emit('agent_interrupt', {
      session_id: sessionId,
      agent_id: agentId,
      token,
    });
  }

  // Typing indicator
  setTyping(sessionId: string, typing: boolean): void {
    if (!this.socket) return;
    this.socket.emit('typing', {
      session_id: sessionId,
      typing,
    });
  }

  // Ping
  ping(): void {
    if (!this.socket) return;
    this.socket.emit('ping');
  }

  // Event subscription
  on(event: string, callback: EventCallback): () => void {
    if (!this.eventCallbacks.has(event)) {
      this.eventCallbacks.set(event, new Set());
    }
    this.eventCallbacks.get(event)!.add(callback);

    // Return unsubscribe function
    return () => {
      this.eventCallbacks.get(event)?.delete(callback);
    };
  }

  // Emit event to all subscribers
  private emit(event: string, data: any): void {
    const callbacks = this.eventCallbacks.get(event);
    if (callbacks) {
      callbacks.forEach((callback) => callback(data));
    }
  }
}

// Singleton instance
export const ws = new WebSocketClient();

File: python/api/v2/websocket.py
"""
WebSocket event handlers for real-time communication
"""

from flask_socketio import emit, join_room, leave_room, disconnect
from flask_jwt_extended import decode_token
from functools import wraps
import json

def authenticated_only(f):
    """Decorator to require authentication for WebSocket events"""
    @wraps(f)
    def wrapped(*args, **kwargs):
        # Get token from query string or handshake
        token = kwargs.get('token')
        if not token:
            emit('error', {'message': 'Authentication required'})
            disconnect()
            return
        
        try:
            # Decode JWT token
            decoded = decode_token(token)
            user_id = decoded['sub']
            kwargs['user_id'] = user_id
            return f(*args, **kwargs)
        except Exception as e:
            emit('error', {'message': 'Invalid token'})
            disconnect()
            return
    
    return wrapped

def register_websocket_events(socketio):
    """Register all WebSocket event handlers"""
    
    @socketio.on('connect')
    def handle_connect(auth):
        """Handle client connection"""
        print(f"Client connected")
        emit('connected', {'status': 'connected'})
    
    @socketio.on('disconnect')
    def handle_disconnect():
        """Handle client disconnection"""
        print("Client disconnected")
    
    @socketio.on('join_session')
    @authenticated_only
    def handle_join_session(data, user_id=None):
        """
        Join a session room for receiving updates
        
        Data:
            - session_id: Session/conversation ID to join
        """
        session_id = data.get('session_id')
        if not session_id:
            emit('error', {'message': 'session_id required'})
            return
        
        # TODO: Verify user has access to this session
        
        join_room(session_id)
        emit('joined_session', {
            'session_id': session_id,
            'message': 'Successfully joined session'
        })
    
    @socketio.on('leave_session')
    def handle_leave_session(data):
        """
        Leave a session room
        
        Data:
            - session_id: Session ID to leave
        """
        session_id = data.get('session_id')
        if not session_id:
            emit('error', {'message': 'session_id required'})
            return
        
        leave_room(session_id)
        emit('left_session', {
            'session_id': session_id,
            'message': 'Successfully left session'
        })
    
    @socketio.on('agent_message')
    @authenticated_only
    def handle_agent_message(data, user_id=None):
        """
        Send message to agent
        
        Data:
            - session_id: Session ID
            - agent_id: Agent ID
            - message: Message content
            - context: Optional context
        """
        from .services.agent_service import AgentService
        
        session_id = data.get('session_id')
        agent_id = data.get('agent_id')
        message = data.get('message')
        context = data.get('context', {})
        
        if not all([session_id, agent_id, message]):
            emit('error', {'message': 'session_id, agent_id, and message required'})
            return
        
        # Join session room if not already joined
        join_room(session_id)
        
        # Emit typing indicator
        emit('agent_typing', {
            'session_id': session_id,
            'agent_id': agent_id,
            'typing': True
        }, room=session_id)
        
        try:
            # Stream agent response
            for chunk in AgentService.execute_streaming(
                agent_id=agent_id,
                user_id=user_id,
                message=message,
                conversation_id=session_id,
                context=context
            ):
                # Emit chunk to room
                emit('agent_response_chunk', {
                    'session_id': session_id,
                    'agent_id': agent_id,
                    'chunk': chunk
                }, room=session_id)
            
            # Emit completion
            emit('agent_response_complete', {
                'session_id': session_id,
                'agent_id': agent_id
            }, room=session_id)
            
        except Exception as e:
            emit('agent_error', {
                'session_id': session_id,
                'agent_id': agent_id,
                'error': str(e)
            }, room=session_id)
        
        finally:
            # Stop typing indicator
            emit('agent_typing', {
                'session_id': session_id,
                'agent_id': agent_id,
                'typing': False
            }, room=session_id)
    
    @socketio.on('agent_interrupt')
    @authenticated_only
    def handle_agent_interrupt(data, user_id=None):
        """
        Interrupt agent execution
        
        Data:
            - session_id: Session ID
            - agent_id: Agent ID
        """
        from .services.agent_service import AgentService
        
        session_id = data.get('session_id')
        agent_id = data.get('agent_id')
        
        if not all([session_id, agent_id]):
            emit('error', {'message': 'session_id and agent_id required'})
            return
        
        # Interrupt agent
        AgentService.interrupt(session_id, agent_id, user_id)
        
        # Emit interruption confirmation
        emit('agent_interrupted', {
            'session_id': session_id,
            'agent_id': agent_id
        }, room=session_id)
    
    @socketio.on('typing')
    def handle_typing(data):
        """
        Broadcast user typing status
        
        Data:
            - session_id: Session ID
            - typing: Boolean
        """
        session_id = data.get('session_id')
        typing = data.get('typing', False)
        
        if not session_id:
            return
        
        emit('user_typing', {
            'session_id': session_id,
            'typing': typing
        }, room=session_id, include_self=False)
    
    @socketio.on('ping')
    def handle_ping():
        """Simple ping/pong for connection health check"""
        emit('pong', {'timestamp': datetime.now().isoformat()})
    
    print("✓ WebSocket events registered")
